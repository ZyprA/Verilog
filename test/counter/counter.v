module counter(clk, reset, q);
    input clk, reset;
    output [3:0] q;
    reg [3:0] q;

    always @(posedge clk or negedge reset)
        if (!reset) q <= 0;
        else q <= q + 1;
endmodule

module counter_sim;
    reg clk, reset;
    wire [3:0] q;

    counter counter_inst(clk, reset, q);

    initial begin
        $dumpfile("counter.vcd");
        $dumpvars(0, counter_sim);
    end
    
    initial begin
        clk = 0;
        forever begin
            #50 clk = ~clk;
        end
    end

    initial begin
        reset = 0;
        #100 reset = 1;
        #100 reset = 0;
        #100 $finish;
    end
endmodule