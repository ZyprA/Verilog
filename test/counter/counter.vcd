$date
	Sun Feb  2 20:06:31 2025
$end
$version
	Icarus Verilog
$end
$timescale
	1s
$end
$scope module counter_sim $end
$var wire 4 ! q [3:0] $end
$var reg 1 " clk $end
$var reg 1 # reset $end
$scope module counter_inst $end
$var wire 1 " clk $end
$var wire 1 # reset $end
$var reg 4 $ q [3:0] $end
$upscope $end
$upscope $end
$enddefinitions $end
$comment Show the parameter values. $end
$dumpall
$end
#0
$dumpvars
b0 $
0#
0"
b0 !
$end
#50
1"
#100
0"
1#
#150
b1 !
b1 $
1"
#200
b0 !
b0 $
0"
0#
#250
1"
#300
0"
