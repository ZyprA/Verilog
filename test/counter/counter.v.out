#! /opt/homebrew/Cellar/icarus-verilog/12.0/bin/vvp
:ivl_version "12.0 (stable)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision + 0;
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/system.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_sys.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_textio.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/v2005_math.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/va_math.vpi";
S_0x156608780 .scope module, "counter_sim" "counter_sim" 2 11;
 .timescale 0 0;
v0x15661cbb0_0 .var "clk", 0 0;
v0x15661cc50_0 .net "q", 3 0, v0x15661ca00_0;  1 drivers
v0x15661cd00_0 .var "reset", 0 0;
S_0x156607520 .scope module, "counter_inst" "counter" 2 15, 2 1 0, S_0x156608780;
 .timescale 0 0;
    .port_info 0 /INPUT 1 "clk";
    .port_info 1 /INPUT 1 "reset";
    .port_info 2 /OUTPUT 4 "q";
v0x1566088f0_0 .net "clk", 0 0, v0x15661cbb0_0;  1 drivers
v0x15661ca00_0 .var "q", 3 0;
v0x15661cab0_0 .net "reset", 0 0, v0x15661cd00_0;  1 drivers
E_0x15660ce50/0 .event negedge, v0x15661cab0_0;
E_0x15660ce50/1 .event posedge, v0x1566088f0_0;
E_0x15660ce50 .event/or E_0x15660ce50/0, E_0x15660ce50/1;
    .scope S_0x156607520;
T_0 ;
    %wait E_0x15660ce50;
    %load/vec4 v0x15661cab0_0;
    %nor/r;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.0, 8;
    %pushi/vec4 0, 0, 4;
    %assign/vec4 v0x15661ca00_0, 0;
    %jmp T_0.1;
T_0.0 ;
    %load/vec4 v0x15661ca00_0;
    %addi 1, 0, 4;
    %assign/vec4 v0x15661ca00_0, 0;
T_0.1 ;
    %jmp T_0;
    .thread T_0;
    .scope S_0x156608780;
T_1 ;
    %vpi_call 2 18 "$dumpfile", "counter.vcd" {0 0 0};
    %vpi_call 2 19 "$dumpvars", 32'sb00000000000000000000000000000000, S_0x156608780 {0 0 0};
    %end;
    .thread T_1;
    .scope S_0x156608780;
T_2 ;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x15661cbb0_0, 0, 1;
T_2.0 ;
    %delay 50, 0;
    %load/vec4 v0x15661cbb0_0;
    %inv;
    %store/vec4 v0x15661cbb0_0, 0, 1;
    %jmp T_2.0;
    %end;
    .thread T_2;
    .scope S_0x156608780;
T_3 ;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x15661cd00_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x15661cd00_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x15661cd00_0, 0, 1;
    %delay 100, 0;
    %vpi_call 2 33 "$finish" {0 0 0};
    %end;
    .thread T_3;
# The file index is used to find the file name in the following table.
:file_names 3;
    "N/A";
    "<interactive>";
    "counter.v";
